psamm/moma.py

```diff
@@ -79,84 +100,332 @@ def __init__(self, model, solver):
         if not result:
             raise MOMAError('Unable to solve initial FBA: {}'.format(
                 result.status))
 
-        return result.get_value(self._v_wt(objective))
-
-    def minimize_l1(self, objective):
-        wt_obj = self._solve_fba(objective)
-
+        # Return the result object
+        return result
+
+    # Returns all the flux values for each of the reactions. This is used in
+    # MOMA LP2 and QLP2 to minimize the change in the flux flow following a
+    # gene deletion.
+    def _get_fba_flux(self, objective):
+        """Returns a dictionary of all the solved fluxes for each reaction
+        in the model."""
+        # Get the result of the wild type FBA problem
+        flux_result = self._solve_fba(objective)
+
+        # Creat a dictionary of all the reaction ids and their fluxes.
+        # {'reaction1': 1000, 'reaction2': 2000, ...}
+        fba_fluxes = {}
+
+        # Place all the flux values in a dictionary
+        for key in self._model.reactions:
+            fba_fluxes[key] = flux_result.get_value(self._v_wt(key))
+
+        # Return the dictionary with all of the fluxes associated with the
+        # reaction id
+        # {'reaction1': 1000, 'reaction2': 2000, ...}
+        return fba_fluxes
+
+    def _get_minimal_flux_fba(self, wt_biomass, objective):
+        """Finds the FBA solution that minimizes all the flux values, while
+        still maximizing the biomass. This still might not find a unique
+        solution."""
+        wt_obj = self._v_wt(objective)
+        with self.constraints(wt_biomass == wt_obj):
+            self._prob.set_objective(self._v_wt.sum(self._all_reactions()))
+            result = self._solve(lp.ObjectiveSense.Minimize)
+        return result
+
+    # Solves for the FBA biomass. Uses the result returned by _solve_fba and
+    # finds the value of the objective (biomass) function.
+    def _get_fba_biomass(self, objective):
+        """Returns the flux for the max biomass found by FBA."""
+        # Run FBA and store the result
+        flux_result = self._solve_fba(objective)
+        # Pull out the biomass value
+        return flux_result.get_value(self._v_wt(objective))
+
+    def minimize_lp2(self, objective, wt_fluxes):
+        """Implemetnation of the lin_MOMA algorithm.
+
+        We try to maximize biomass, by keeping the fluxes relitively the same
+        as the wildtype. This helps avoid the assumption that an organism will
+        perform optimally directly after removing a gene.
+
+        This might not be the optimal minimization solution for the wild type
+        and the knockout vectors.
+
+        Minimizes sum of |wild type - knockout|
+
+        Args:
+            objective: Biomass reaction for the model.
+            wt_fluxes: All the wt_fluxes found by FBA in the form of a
+                dictionary. Use :meth: _get_fba_flux(objective) to return
+                a dictionary.
```

Comment:
I think it's great to have a recommendation here specifying how this input value can be obtained. However, the `_get_fba_flux()` method is private (because of the initial `_`) so this contradicts the message. Either, we should make it public by dropping the initial underscore or add a different recommendation that is public (I think making the method public makes sense). Also, keep in mind that the paper on MOMA suggests that the fluxes could also come from experimental results, i.e. the dictionary is not required to come from FBA it is only a suggestion. We may want to move the recommendation to use the FBA fluxes to a paragraph above and then just specify that the argument should be a dictionary of flux values.